### PR TITLE
Make fake int mod op (%) actually return int

### LINKED ISSUE #35152: Shader language GLES2 type issue with mod
**Godot version:**
Godot_v3.2-beta6_x11.64

**OS/device including version:**
Kubuntu 19.10

**Issue description:**
Running the code below (the second line) with GLES2 gives the following error in the output tab, the shader editor is unaware, and there some glitch drawing. Wrapping the expression in an int() works around it. It works in GLES3.
```
 drivers/gles2/shader_gles2.cpp:129 - SceneShaderGLES2: Fragment shader compilation failed:
0:1274(2): error: initializer of type float cannot be assigned to variable of type int

 drivers/gles2/shader_gles2.cpp:360 - Method/Function Failed, returning: __null
 drivers/gles2/shader_gles2.cpp:88 - Condition ' !version ' is true. returned: false
 drivers/gles2/shader_gles2.h:257 - Condition ' !version ' is true. returned: -1
 drivers/gles2/shader_gles2.h:257 - Condition ' !version ' is true. returned: -1
```

**Steps to reproduce:**
```
vec2 fragcord = UV * 10.0;
int val = int(fragcord.x) % 4;
```

### PR DESCRIPTION
**Pull request by xix-xeaon**
Fixes #35152 
Using this opportunity as a learning experience, I really hope I didn't manage to screw up with such a small change =P I compiled and tested it though, works fine.

### CODE REVIEW

**Comment by bojidar-bg** [comment]
Does it work the same as GLES3 with negative numbers (on either side of the operator)?

File: drivers/gles2/shader_compiler_gles2.cpp [hunk 1]
```diff
@@ -755,11 +755,11 @@ String ShaderCompilerGLES2::_dump_node_code(SL::Node *p_node, int p_level, Gener
 
 				case SL::OP_MOD: {
 
-					code += "mod(float(";
+					code += "int(floor(0.5+mod(float(";
```
Comment (akien-mga) [inline drivers/gles2/shader_compiler_gles2.cpp]: I think the `floor()` is unnecessary, `int()` will already truncate. Please also add spaces around `+` to use the same style as for our C++ code (even though we can't run `clang-format` on this generated code).

Comment (xix-xeaon) [reply drivers/gles2/shader_compiler_gles2.cpp]: Right, makes sense.

**Comment by xix-xeaon** [comment]
@bojidar-bg It seems it does not. This only fixes the type issue.

**Comment by xix-xeaon** [comment]
@akien-mga I can fix the sign too, but I suppose it should be a different pull?

**Comment by xix-xeaon** [comment]
Hmm it seems my fix for the sign behavior ended up here too now. Well, let me know if I should do something differently.

File: drivers/gles2/shader_compiler_gles2.cpp [hunk 2]
```diff
@@ -754,12 +754,21 @@ String ShaderCompilerGLES2::_dump_node_code(SL::Node *p_node, int p_level, Gener
...
 
 				case SL::OP_MOD: {
-
-					code += "mod(float(";
+					code += "(";
 					code += _dump_node_code(op_node->arguments[0], p_level, r_gen_code, p_actions, p_default_actions, p_assigning);
-					code += "), float(";
+					code += " < 0 ? ";
+					code += "-int(0.5 + mod(float(-";
+					code += _dump_node_code(op_node->arguments[0], p_level, r_gen_code, p_actions, p_default_actions, p_assigning);
```
Comment (akien-mga) [inline drivers/gles2/shader_compiler_gles2.cpp]: The ternary operator business might be more readable if you save the output of the `_dump_node_code` calls for both argument 0 and 1 in variables, and use those to construct `code`.

Comment (akien-mga) [reply drivers/gles2/shader_compiler_gles2.cpp]: Could you also squash the commits together eventually?

**Comment by nathanwfranke** [comment]
I get this with the PR

```
ERROR: _display_error_with_code: CanvasShaderGLES2: Fragment shader compilation failed:
0(592) : error C1101: ambiguous overloaded function reference "abs(int)"
    (0) : gp5 cp50 fp50 vp50 gp50 cpf fpf vpf gpf superp float abs(superp float)
    (0) : ps gp5 gp4 cp50 fp50 vp50 gp50 cpf fpf vpf gpf lowp float abs(lowp float)
    (0) : ps gp5 gp4 cp50 fp50 vp50 gp50 cpf fpf vpf gpf mediump float abs(mediump float)
    (0) : float abs(float)
0(592) : error C1101: ambiguous overloaded function reference "abs(int)"
    (0) : gp5 cp50 fp50 vp50 gp50 cpf fpf vpf gpf superp float abs(superp float)
    (0) : ps gp5 gp4 cp50 fp50 vp50 gp50 cpf fpf vpf gpf lowp float abs(lowp float)
    (0) : ps gp5 gp4 cp50 fp50 vp50 gp50 cpf fpf vpf gpf mediump float abs(mediump float)
    (0) : float abs(float)

   At: drivers/gles2/shader_gles2.cpp:129.
ERROR: get_current_version: Method/Function Failed, returning: __null
   At: drivers/gles2/shader_gles2.cpp:360.
ERROR: bind: Condition ' !version ' is true. returned: false
   At: drivers/gles2/shader_gles2.cpp:88.
```

Code:
```
shader_type canvas_item;

void fragment() {
	int i = 5 % 2;
}
```

File: drivers/gles2/shader_compiler_gles2.cpp (hunk 2 above)
Comment (xix-xeaon) [reply drivers/gles2/shader_compiler_gles2.cpp]: I realized I was being silly building on the original implementation anyway so I replaced the whole thing. However, I can't figure out how to do the squashing; the only thing I've found is the "Squash and Merge" option for a PR but I don't have permission to merge here obviously.

Comment (akien-mga) [reply drivers/gles2/shader_compiler_gles2.cpp]: See our how-to on https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#mastering-the-pr-workflow-the-rebase

**Comment by xix-xeaon** [comment]
@nathanwfranke I have no idea about that and I never got that error, but my new implementation doesn't use abs. Try just using abs with a few different values and if you get the error report a new issue.

**Comment by nathanwfranke** [comment]
I'm not using abs, just `int i = 5 % 2`. Works fine in beta 6, I can check master

**Comment by xix-xeaon** [comment]
Nah, this PR used abs when you tested it, and it seems that's where your error is from. It does not use abs anymore though.

**Comment by nathanwfranke** [comment]
@xix-xeaon Thank you, it seems to work now. Possibly was my fault because I am quite new to this whole pull request thing :smile:

File: drivers/gles2/shader_compiler_gles2.cpp [hunk 3]
```diff
@@ -754,12 +754,9 @@ String ShaderCompilerGLES2::_dump_node_code(SL::Node *p_node, int p_level, Gener
...
 				case SL::OP_MOD: {
-
-					code += "mod(float(";
-					code += _dump_node_code(op_node->arguments[0], p_level, r_gen_code, p_actions, p_default_actions, p_assigning);
-					code += "), float(";
-					code += _dump_node_code(op_node->arguments[1], p_level, r_gen_code, p_actions, p_default_actions, p_assigning);
-					code += "))";
+					String a = _dump_node_code(op_node->arguments[0], p_level, r_gen_code, p_actions, p_default_actions, p_assigning);
+					String n = _dump_node_code(op_node->arguments[1], p_level, r_gen_code, p_actions, p_default_actions, p_assigning);
+					code += "(" + a + " - " + n + " * (" + a + " / " + n + "))";
```
Comment (akien-mga) [inline drivers/gles2/shader_compiler_gles2.cpp]: How does it behave when `n == 0`?

Comment (bojidar-bg) [reply drivers/gles2/shader_compiler_gles2.cpp]: In C/C++ integer modulo with 0 results in SIGFPE. So, undefined behavior?

Comment (akien-mga) [inline drivers/gles2/shader_compiler_gles2.cpp]: Should probably add a `floor()` in there to match the official spec: https://www.khronos.org/registry/OpenGL-Refpages/gl4/html/mod.xhtml

Comment (akien-mga) [reply drivers/gles2/shader_compiler_gles2.cpp]: Well `floor(int)` won't work, but here the mod implementation will be wrong for float values.

Comment (xix-xeaon) [reply drivers/gles2/shader_compiler_gles2.cpp]: On my computer GLES3 a % 0 always gives 0. This PR always gives a.

Comment (xix-xeaon) [reply drivers/gles2/shader_compiler_gles2.cpp]: There'd be no point, right? As far as I understand this operation already require both values to be int.

Comment (akien-mga) [reply drivers/gles2/shader_compiler_gles2.cpp]: Correct, my bad.

BTW, the same changes are needed for `OP_ASSIGN_MOD` (`i %= n`).

**Comment by akien-mga** [comment]
Code seems good to me, but given that we're already at RC 2 for Godot 3.2, I'll leave it for after the release, to be cherry-picked after testing for 3.2.1.

**Comment by clayjohn** [comment]
@xix-xeaon Sorry that this one slipped through the cracks. Do you mind rebasing it so it is ready to be cherrypicked?

**Comment by xix-xeaon** [comment]
@clayjohn Alright, I think it's good.

**Review by clayjohn — APPROVED** [review]

**Comment by akien-mga** [comment]
This should be rebased to target the `3.2` branch as there is no GLES2 in `master` anymore.

**Comment by nathanfranke** [comment]
Was GLES2 Completely removed from master? Isn't it just disabled while it's unstable?

**Comment by akien-mga** [comment]
> Was GLES2 Completely removed from master? Isn't it just disabled while it's unstable?

It's not unstable, it's unimplemented. That was dead code that doesn't compile, and will require a full rewrite and not just some build fixes, so I removed it.

**Comment by akien-mga** [comment]
I force pushed a commit rebased to target the `3.2` branch myself.

Note that it means your fork's `master` branch is now `3.2` + this commit, you'll have to `git reset --hard` it to the upstream `master` branch if you want to work on `master` in the future.

**Comment by akien-mga** [comment]
Thanks!

**Comment by lawnjelly** [comment]
Is this doing a runtime check for zero? If so I was just wondering, is it actually necessary to do this check? Afaik shaders are required to work gracefully even with things like divide by zero and pass INF. Any use of % 0 is an error anyway, so I'm wondering whether it is worth sacrificing performance for this. Has it been performance tested with and without the check?

**Comment by nathanfranke** [comment]
@lawnjelly Can't say for sure, but this appears to just transpile to GLSL and I doubt GLSL would have a zero check.

**Comment by akien-mga** [comment]
> Is this doing a runtime check for zero? If so I was just wondering, is it actually necessary to do this check? Afaik shaders are required to work gracefully even with things like divide by zero and pass INF. Any use of % 0 is an error anyway, so I'm wondering whether it is worth sacrificing performance for this. Has it been performance tested with and without the check?

@lawnjelly See https://github.com/godotengine/godot/issues/35152#issuecomment-574822802 for the rationale.

CC @clayjohn @Chaosus

**Comment by lawnjelly** [comment]
Same will apply with mod function, afaik it is a requirement for shaders to run without problem even with divide by zero and other exceptions (the result may be undefined but they should run).

**Pull request merged**